### DISCUSSION TITLE
texture from dynamically generated SVG in pixiJS v8

### DISCUSSION THREAD

**Discussion by ripmurdock** [pixijs/pixijs#10953] [Q&A]
Hi,

How can I create a texture from a dynamically generated SVG in pixiJS v8?

This worked in v7:
const testSVG = '<svg viewBox="0 0 100 100" width="7" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="50" /></svg>'
, texture = PIXI.Texture.from(testSVG)

but texture logs out as undefined in v8.

And this approach:

  const textureSource = PIXI.TextureSource.from(testsvg )
  const texture = PIXI.Texture.from(textureSource)

Throws this error:
Error: Could not find a source type for resource: <svg viewBox="0 0 100 100" width="7" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="50" /></svg>

Thanks!

**Comment by ripmurdock** [comment]
I was able to get pixiJS v8 to output an svg with this multi-step process:

    const renderer = await PIXI.autoDetectRenderer({
      preference,
      clearBeforeRender: true,
      backgroundAlpha: 0,
      backgroundColor: 0x000000,
      width: 500,
      height: 500
      resolution: 1,
      antialias: false,
      hello: true
    });
    
    testsvg = '<SVG viewBox="0 0 100 100" fill="red" width="200" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="50" /></SVG>'
    
    const graphics = new PIXI.Graphics().svg(testsvg)
      
    , texture = renderer.generateTexture(graphics).
    
    , sprite = new PIXI.Sprite(texture)
    
    ;


I'd still be curious if I'm missing a more straightforward way of outputting an svg in v8, similar to v7's one liner.

**Reply by bhousel** [reply]
With v8 we ended up going back to an older way of doing this now that [`SVGResource`](https://pixijs.download/v7.x/docs/PIXI.SVGResource.html) is gone..
You can put the svg into an `HTMLImageElement` and then create the texture something like this:

```js
    let image = new Image();
    image.src = `data:image/svg+xml,${encodeURIComponent(svgString)}`;
    image.onload = () => {
      const texture = PIXI.Texture.from(image);
      // save texture and release references as needed?
      // image.onload = null;
      // image = null;
    };
```

Some caveats:  It introduces some async, and you shouldn't do a lot (thousands?) of these.  I think the browser will queue this work as microtasks.  (I think `PIXI.SVGResource` worked this way anyway)

**Reply by ripmurdock** [reply]
Thanks.

I'm animating about 5000 sprites.  I've been able to get this to work well in a v7 ParticleContainer.

Are there settings that would allow a v8 Container to render as fast as a v7 ParticleContainer?

Or is a ParticleContainer planned for v8?

Thanks again!

**Reply by bigtimebuddy** [reply]
@ripmurdock https://pixijs.com/blog/particlecontainer-v8

**Comment by ripmurdock** [comment]
Wow. Thanks!

Now I'm wondering if I should convert my svg texture based v7 ParticleContainers to v8 ParticleContainers.

Will I be able to add more particles and get the same speed if the textures are created from svg's and all of the properties are made dynamic?

Also, is it necessary to put an SVG into an HTMLImageElement element in v8, if the svg is created dynamically and does not have to be loaded?

In v7, I created textures for particles directly from a string, like this:

const svg = '<svg viewBox="0 0 800 800" width="136" fill="red" xmlns="http://www.w3.org/2000/svg"><circle cx="400" cy="400" r="400" /></svg>'
,texture = PIXI.Texture.from(svg);

Is either of these v8 approaches to creating a texture from an svg string faster, new Image to PIXI.Texture.from or  PIXI.Graphics().svg to generateTexture()?

HTMLImageElement based:
    let image = new Image();
    image.src = `data:image/svg+xml,${encodeURIComponent(svgString)}`;
    image.onload = () => {
      const texture = PIXI.Texture.from(image);
      // save texture and release references as needed?
      // image.onload = null;
      // image = null;
    };

Or PIXI.Graphics based
const graphics = new PIXI.Graphics().svg(svg)
, texture = renderer.generateTexture(graphics).

Thanks again.

**Reply by bhousel** [reply]
> Also, is it necessary to put an SVG into an HTMLImageElement element in v8, if the svg is created dynamically and does not have to be loaded?

The only reason we are doing things like this is because we have a lot of SVG icons that we use outside of Pixi in other parts of the document.

If you're creating your SVGs dynamically, it will be easier for you to just create Pixi Graphics dynamically instead.

**Reply by GitHubStig** [reply]
@ripmurdock
Suggestion: instead of using `image.onload = ...`, consider using `await image.decode()` instead.
```
const image = new Image();
image.src = `data:image/svg+xml,${encodeURIComponent(svgString)}`;
await image.decode()
const texture = PIXI.Texture.from(image);
...
```
reference: https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/decode

**Reply by ripmurdock** [reply]
Thanks.

How can I scale an svg that has a viewBox for use in a v8 particle container?

I've tried the svg width with and without units, and I've tried setting the width property on the graphics, texture, and particle objects.

The particle ends up with the width of the viewBox, not scaled to the width of the svg:

testsvg = '<SVG viewBox="0 0 800 800" fill="red" width="100" xmlns="http://www.w3.org/2000/svg"><circle cx="400" cy="400" r="400" /></SVG>'

const graphics = new PIXI.Graphics().svg(testsvg)
, texture = renderer.generateTexture(graphics)
, particle = new PIXI.Particle(texture)

Thanks again.

**Comment by ripmurdock** [comment]
I haven't been able to successfully apply a width and height to either a texture or a particle in v8.

testsvg = '<SVG viewBox="0 0 800 800" fill="red" width="100px" xmlns="http://www.w3.org/2000/svg"><circle cx="400" cy="400" r="400" /></SVG>'

const graphicsContext5 = new PIXI.GraphicsContext().svg(testsvg) 
const graphics5 = new PIXI.Graphics(graphicsContext5)
graphics5.width = 10
graphics5.height = 10

const texture5 = renderer.generateTexture(graphics5)
texture5.width = 10
texture5.height = 10

const particle5 = new PIXI.Particle(texture5)
const sprite5 = new PIXI.Sprite(texture5)
console.log("sprite5 initial width",sprite5.width)
sprite5.width = 10
sprite5.height = 10

console.log("graphics5 width",graphics5.width)
console.log("texture5 width",texture5.width)
console.log("particle5 width",particle5.width)
console.log("sprite5 width",sprite5.width)

//output
sprite5 initial width 800
graphics5 width 10
texture5 width 800
particle5 width undefined
sprite5 width 10

texture5 has a width equal to the viewBox width, and width is undefined for the v8 particle object.

I also haven't been able to figure out where the width and height are defined in the provided v8 particle bunny example:

      BunnyParticleV8 = class {
        constructor(texture, bounds) {
          this.gravity = 0.2;
          this.speedX = Math.random() * 10;
          this.speedY = Math.random() * 10 - 5;
          this.positionX = 0;
          this.positionY = 0;
          this.view = new Particle({
            texture,
            anchorX: 0.5,
            anchorY: 1
          });
          this.bounds = bounds;
        }
        update() {
          let pX = this.positionX;
          let pY = this.positionY;
          pX += this.speedX;
          pY += this.speedY;
          this.speedY += this.gravity;
          if (pX > this.bounds.right) {
            this.speedX *= -1;
            pX = this.bounds.right;
          } else if (pX < this.bounds.left) {
            this.speedX *= -1;
            pX = this.bounds.left;
          }
          if (pY > this.bounds.bottom) {
            this.speedY *= -0.85;
            pY = this.bounds.bottom;
            if (Math.random() > 0.5) {
              this.speedY -= Math.random() * 6;
            }
          } else if (pY < this.bounds.top) {
            this.speedY = 0;
            pY = this.bounds.top;
          }
          this.view.x = this.positionX = pX;
          this.view.y = this.positionY = pY;
        }
        reset() {
          this.positionX = 0;
          this.positionY = 0;
        }
      };

What's the correct way in v8 to define the width and height of a particle object or a texture used to create a particle object?

Thanks!
